### PR TITLE
WT-6752 Assert history store updates are all committed

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/history/hs_rec.c [hunk 1]
```diff
@@ -669,6 +669,26 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
+            /*
+             * Skip the check if we don't have a snapshot. Therefore, the check doesn't ensure that
+             * we never write uncommitted updates to the history store. But this is the best we can
+             * do for now.
+             *
+             * Sometimes the application threads and the checkpoint thread will fall behind the
+             * eviction threads, it may choose an invisible update to write to the disk if a failed
+             * eviction decided to write that update to the disk before. In this case, no point to
+             * check the visibility of the history store updates.
+             */
```
Comment (hbokhari) [inline src/history/hs_rec.c:686]: Can you reword it to something like this:

Sometimes the application and the checkpoint threads will fall behind the eviction threads, and they will choose an invisible update that was previously selected by an eviction pass. In that case, there is no point in checking the visibility of such history store updates.

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -669,6 +669,26 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
+            WT_ASSERT(session,
```
Comment (hbokhari) [inline src/history/hs_rec.c:687]: We normally use `conn->txn_global->last_running` to check visibility of updates when there are no snapshots. We can use the same test here.

Comment (quchenhao) [reply src/history/hs_rec.c:687]: No, we can't do that here. Some of the invisible updates will be selected because of the special case I described above.

File: src/history/hs_rec.c (hunk 1 above)
Comment (quchenhao) [reply src/history/hs_rec.c:686]: reworded

File: src/history/hs_rec.c (hunk 2 above)
Comment (hbokhari) [reply src/history/hs_rec.c:687]: I am not sure how useful this check is here then. By the time eviction threads reach here, they would have already released the snapshot. Not sure though, but is it possible to move these checks up somewhere we have already the snapshot for eviction threads? I believe most of the reconcile work is either performed by checkpoint or eviction threads so we kind of miss almost half of the scenarios.

**Comment by quchenhao** [comment]
@hbokhari If eviction has a snapshot, it is not released at this point. We are still in wt_recouncile.

**Comment by hbokhari** [comment]
@quchenhao - Thank you for pointing it out. I confused this with something else.

**Review by hbokhari — APPROVED** [review]
LGTM!

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -669,6 +669,26 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
+             * we never write uncommitted updates to the history store. But this is the best we can
+             * do for now.
+             *
+             * Sometimes the application and the checkpoint threads will fall behind the eviction
+             * threads, and they may choose an invisible update if the update was previously
+             * selected by a failed eviction pass. In that case, there is no point in checking the
+             * visibility of such history store updates.
+             */
+            WT_ASSERT(session,
+              !F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT) ||
```
Comment (bvpvamsikrishna) [inline src/history/hs_rec.c:688]: can we check this case irrespective of whether transaction has a snapshot or not. In other words, can we have a utility function which returns whether a transaction is committed or not irrespective of caller transaction state and use that here.

Comment (quchenhao) [reply src/history/hs_rec.c:688]: I'm not sure if that's viable and I'll have a try.

File: src/txn/txn.c [hunk 4]
```diff
@@ -138,6 +138,50 @@ __wt_txn_release_snapshot(WT_SESSION_IMPL *session)
...
+/*
+ * __wt_txn_in_global_list --
+ *     Check if a transaction is still in the global transaction list. If it is not, then it is
+ *     committed, prepared, or rolled back. It is possible that we race with commit, prepare or
+ *     rollback and a transaction is still in the global list before the start of the call is
+ *     reported not in the list.
+ */
+bool
+__wt_txn_in_global_list(WT_SESSION_IMPL *session, uint64_t txnid)
+{
```
Comment (quchenhao) [inline src/txn/txn.c:150]: I named it as __wt_txn_in_global_list because it is not a reliable check. Name it as __wt_txn_uncommitted would be misleading.

File: src/txn/txn.c [hunk 5]
```diff
@@ -138,6 +138,50 @@ __wt_txn_release_snapshot(WT_SESSION_IMPL *session)
     }
 }
 
+/*
+ * __wt_txn_in_global_list --
+ *     Check if a transaction is still in the global transaction list. If it is not, then it is
+ *     committed, prepared, or rolled back. It is possible that we race with commit, prepare or
+ *     rollback and a transaction is still in the global list before the start of the call is
+ *     reported not in the list.
```
Comment (hbokhari) [inline src/txn/txn.c:146]: and a transaction is still in the global list before the start of the call is reported not in the list ->
, and a transaction can be in the global list before the start of the call but eventually reported as not being in the list.

File: src/txn/txn.c [hunk 6]
```diff
@@ -138,6 +138,50 @@ __wt_txn_release_snapshot(WT_SESSION_IMPL *session)
...
+
+    conn = S2C(session);
+    txn_global = &conn->txn_global;
+    in_global_list = true;
+
+    /* We're going to scan the table: wait for the lock. */
+    __wt_readlock(session, &txn_global->rwlock);
+    oldest_id = txn_global->oldest_id;
+
+    if (txnid <= oldest_id) {
```
Comment (hbokhari) [inline src/txn/txn.c:166]: Just double checking that this should be "<=" and not "<"? If the `txn_id` is the current `oldest_id` then we should return true.

**Review by hbokhari — APPROVED** [review]
I have few minor comments but otherwise LGTM.

File: src/txn/txn.c (hunk 6 above)
Comment (quchenhao) [reply src/txn/txn.c:166]: Thanks for the comment. "<=" is correct, but it makes question the naming of the function again. Here we are checking whether the txn_id is behind the oldest id not against the ongoing transaction list.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

File: src/txn/txn.c (hunk 6 above)
Comment (quchenhao) [reply src/txn/txn.c]: Sorry, you are right here we should use "<" instead as last_running may not be committed.
